pkg/signaling/mcast/backend.go

```diff
@@ -0,0 +1,141 @@
+// SPDX-FileCopyrightText: 2025 Adam Rizkalla <ajarizzo@gmail.com>
+// SPDX-License-Identifier: Apache-2.0
+
+package mcast
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"syscall"
+
+	"cunicu.li/cunicu/pkg/crypto"
+	"cunicu.li/cunicu/pkg/log"
+	signalingproto "cunicu.li/cunicu/pkg/proto/signaling"
+	"cunicu.li/cunicu/pkg/signaling"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/proto"
+)
+
+func init() { //nolint:gochecknoinits
+	signaling.Backends["multicast"] = &signaling.BackendPlugin{
+		New:         NewBackend,
+		Description: "Multicast",
+	}
+}
+
+type Backend struct {
+	signaling.SubscriptionsRegistry
+
+	conn       *net.UDPConn
+	mcast_addr *net.UDPAddr
+	config     BackendConfig
+
+	logger *log.Logger
+}
+
+func NewBackend(cfg *signaling.BackendConfig, logger *log.Logger) (signaling.Backend, error) {
+	b := &Backend{
+		SubscriptionsRegistry: signaling.NewSubscriptionsRegistry(),
+		logger:                logger,
+	}
+
+	var err error
+
+	if err = b.config.Parse(cfg); err != nil {
+		return nil, fmt.Errorf("failed to parse backend configuration: %w", err)
+	}
+
+	// Parse multicast group address
+	if b.mcast_addr, err = net.ResolveUDPAddr("udp", b.config.Target); err != nil {
+		return nil, fmt.Errorf("Error parsing multicast address: %w", err)
+	}
+
+	// Add listener for multicast group
+	if b.conn, err = net.ListenMulticastUDP("udp", b.config.Options.Interface, b.mcast_addr); err != nil {
+		return nil, fmt.Errorf("Error adding multicast listener: %w", err)
+	}
+
+	if b.config.Options.Loopback {
+		// Enable multicast loopback
+		fd, _ := b.conn.File()
+		syscall.SetsockoptInt(int(fd.Fd()), syscall.IPPROTO_IP, syscall.IP_MULTICAST_LOOP, 1)
+	}
+
+	go func() {
+		buf := make([]byte, 4096)
+		for {
+			n, err := b.conn.Read(buf)
+			if err != nil {
+				if err == net.ErrClosed {
+					break
+				}
+				b.logger.Error("Error reading from UDPConn", zap.Error(err))
+				continue
+			}
+
+			var env signalingproto.Envelope
+			err = proto.Unmarshal(buf[:n], &env)
+			if err != nil {
```

Comment:
For concsiseness, I prefer to call the function and check its return value in a single line:

```suggestion
			if err = proto.Unmarshal(buf[:n], &env); err != nil {
```